### ISSUE TITLE
[v0.3+] perf: Optimized Softmax kernel (xFormers/AITemplate-style)

### ISSUE THREAD

**Issue by m96-chan** [m96-chan/PyGPUkit#79]
## Summary

Optimize the softmax CUDA kernel using techniques from xFormers, AITemplate, and llm.c.

## Current Implementation

Basic warp-level reduction in `native/ops/nn/nn_kernels.cuh`:
- Single-precision reads/writes
- Simple block-level reduction via shared memory
- Numerically stable (max subtraction)

### 1. Vectorized Memory Access
Use packed types (`float4`, `half4`) for 4-8 elements per load/store.

**Expected Gain**: 2-4x memory bandwidth improvement

```cpp
// Current
float val = input[idx];

// Optimized
float4 val = reinterpret_cast<const float4*>(input)[idx / 4];
```

### 2. Warp-Only Reduction (Small K)

For K < 1920 (K/pack_size < 32), use warp-only reduction with `__shfl_xor_sync()`:

```cpp
// Warp reduction without shared memory
for (int offset = 16; offset > 0; offset /= 2) {
    max_val = fmaxf(max_val, __shfl_xor_sync(0xffffffff, max_val, offset));
}
```

**Threshold**: Use warp-only when `K / pack_size < 32`

### 3. Online Softmax Algorithm

Merge max and sum into single pass (reduces memory round-trips):

```cpp
// Traditional: 3 passes (max, exp+sum, normalize)
// Online: 2 passes (max+sum fused, normalize)
float m_new = fmaxf(m_prev, x);
float exp_diff = expf(m_prev - m_new);
d_new = d_prev * exp_diff + expf(x - m_new);
```

Reference: [Online normalizer calculation for softmax](https://arxiv.org/abs/2001.04438)

### 4. Register Blocking with Unrolling

```cpp
constexpr int UNROLL = 8;
float local_max[UNROLL];
#pragma unroll
for (int i = 0; i < UNROLL; i++) {
    local_max[i] = __ldcs(&input[base + i * stride]);
}
```

### 5. Cache-Bypassing Instructions

For large K, use `__ldcs` (load cached streaming) to avoid L1 pollution:

```cpp
float val = __ldcs(&input[idx]);  // L2 only, bypass L1
```

## Performance Targets

| Metric | Current | Target |
|--------|---------|--------|
| GB/s (RTX 3090 Ti) | ~300 | 800+ |
| Efficiency vs peak BW | ~30% | 80%+ |

Reference: [CudaSoftmax](https://github.com/fattorib/CudaSoftmax) achieves 452.7 GB/s on RTX 4070 (90% of peak).

## Implementation Plan

1. [ ] Add float4 vectorized loads/stores
2. [ ] Implement warp-only path for small K
3. [ ] Add online softmax algorithm
4. [ ] Register blocking with unroll factor
5. [ ] Benchmark and tune thresholds

## References

- [AITemplate Wiki: How to write a fast Softmax CUDA kernel?](https://github.com/facebookincubator/AITemplate/wiki/How-to-write-a-fast-Softmax-CUDA-kernel%3F)
- [llm.c softmax_forward.cu](https://github.com/karpathy/llm.c/blob/master/dev/cuda/softmax_forward.cu)
- [OneFlow: How to Implement an Efficient Softmax CUDA Kernel](https://oneflow2020.medium.com/how-to-implement-an-efficient-softmax-cuda-kernel-oneflow-performance-optimization-sharing-405ad56e9031)
- [CudaSoftmax](https://github.com/fattorib/CudaSoftmax)
- [PyTorch PersistentSoftmax.cuh](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/PersistentSoftmax.cuh)
